Update bambu_virtual_printer.py

CodeMasterCody3D opened the pull request:
Problem 1:fixed

Scenario of problem:When you cancel a print on the printers screen or on Bambu handy app, you can't start another print from orcaslicer. -FIXED

Problem 2:fixed

Scenario of problem:when selecting SD card file (A) then selection of local file, going back to SD card file (A) will not work, but going to SD card file (B) for selection does work, then you can successfully switch back to SD card (A). No one wants to go through that. -FIXED

Problem 3:fixed

Scenario of problem:When you cancel a print that was started via octoprint or orcaslicer, you will not be able to start another print with orcaslicer.-FIXED

CodeMasterCody3D committed:
Update bambu_virtual_printer.py

Problem 1:fixed

Scenarioof problem:When you cancel a print on the printers screen or on Bambu handy app, you can't start another print from orcaslicer. -FIXED

Problem 2:fixed

Scenario of problem:when selecting SD card file (A) then selection of local file, going back to SD card file (A) will not work, but going to SD card file (B) for selection does work, then you can successfully switch back to SD card (A). No one wants to go through that. -FIXED

Problem 3:fixed

Scenario of problem:When you cancel a print that was started via octoprint or orcaslicer, you will not be able to start another print with orcaslicer.-FIXED

jneilliii commented on octoprint_bambu_printer/printer/bambu_virtual_printer.py:
```diff
@@ -547,18 +644,41 @@ def speed_adjust(speed_percentage):
             if self.bambu_client.publish(GCODE_COMMAND):
-                self._log.info("command sent successfully")
-                self.sendOk()
+                self._log.info(f"command {gcode} sent successfully via MQTT. Sending ok {next_expected_line}")
```
prefer this logging to be debug instead of info.

CodeMasterCody3D replied:
Thank you for the comment. I'll be fixing that very soon. I'm still new to all this but it's coming around slowly.

jneilliii replied:
still not debug.

jneilliii commented on octoprint_bambu_printer/printer/bambu_virtual_printer.py:
```diff
@@ -352,23 +359,70 @@ def select_project_file(self, file_path: str) -> bool:
 
+    ##~~ custom command for debugging
+
+    @gcode_executor.register_no_data("M999")
```
I'm hesitant to introduce custom gcode commands. You're already using the [select_file](https://docs.octoprint.org/en/master/modules/printer.html#octoprint.printer.PrinterInterface.select_file) from the virtual printer interface, wonder if we could utilize [unselect_file](https://docs.octoprint.org/en/master/modules/printer.html#octoprint.printer.PrinterInterface.unselect_file) in its place?

CodeMasterCody3D replied:
I may have forgotten to take that out. Bunch of testing was going on and it's been a long night. I think that select_file isn't actually needed and that was just apart of my testing i forgot to remove. I will get back to you soon when I find out if select_file is needed or not.

CodeMasterCody3D replied:
Also I believe I can remove the m999 custom command.

CodeMasterCody3D committed:
Update bambu_virtual_printer.py

##didnt need the select_file, also the my m999 was implemented into the M23 command already and the m999 was a debugging test.
I also changed the sleep times as well to 0.05

CodeMasterCody3D reviewed (commented):
Fixed

jneilliii commented on octoprint_bambu_printer/printer/bambu_virtual_printer.py:
```diff
@@ -359,16 +366,46 @@ def _sd_status(self) -> bool:
 
-        self.sendIO(f"File opened: {self.selected_file.dosname} Size: {self.selected_file.size}")
+        self.sendIO(
+            f"File opened: {self.selected_file.file_name}  "
```
Sorry if I missed this before, but I'm pretty sure that using file_name here instead of dosname is what causes OctoPrint to not recognize the file that is being printed currently. Please explain the reasoning behind changing this?

jneilliii commented on octoprint_bambu_printer/printer/bambu_virtual_printer.py:
```diff
@@ -703,6 +765,12 @@ def _trigger_change_state(self, new_state: APrinterState):
 
         self._current_state.finalize()
         self._current_state = new_state
+
+        # Check if the new state is the IdleState (self._state_idle is the instance of IdleState)
+        if new_state == self._state_idle:
+            self._log.debug("Transitioned to Idle state. Applying cleanup delay and removing selection.")
+
+
```
Was this also just your debugging method? Don't see a reason for logging this.

CodeMasterCody3D replied:
Thank you for pointing this out. Yes I am working on the AMS as well right now. I can't get it to print with the AMS if starting print with octoprint I was messing around and trying to get the exact JSON message it sends to the printer to see how it's structuring the ams_mapping when i start a print. 

Since I'm new it's hard to debug things i can't see. 

I am on printer version 01.08.00.00 X1plus
My ams has updated to latest. Is this bad?

Should I use a different version of my printer firmware?  I can load my old version if needed.

jneilliii replied:
I have no idea on firmware and what happens outside of X1+ firmware. I only have an X1C running that firmware. The AMS mapping is very difficult and convoluted and I know exactly how it works. I'd recommend you not work on that bit of code, because of the great complication involved and you will be duplicating efforts of what I've been working on. The plugin currently will work with AMS mapping if you start the print from OctoPrint's web interface for an SD card stored file. There is another issue around here that describes how this works.

CodeMasterCody3D replied:
I fixed the AMS.

The first problem was that the plugin couldn't determine if I had ams#2  or ams#1 plugged in by themselves or not. 


I had my ams#2 plugged in the whole time by it's self, that's why I never could get ams working before. 

I did tons of research,  through all the research and development I didn't find out about the ams being hard coded till yesterday. 


Thankfully,  with the debugging I set, I was able to see the json mqtt print command I was sending over to the printer it was correct the whole time but since I had ams#2 plugged in, it was sending print commands as if it was ams#1

We need a way to determine if it's ams#1 or 2.

From all the research and development I did, I have the ams now able to print in ANY SCENARIO!! Only thing i haven't added is manual override with in the UI.

When you you send a print from orcaslicer, it will extract both, the Metadata/plate_1.json and the Metadata/slice_info.config 

For the slice_info.config it parses the xml into a json format, writes it to a file and names it after the print file. Fortunately we don't need to use the 8.3 name for this part specifically. I thought it would have so the code I first made had 2 copies 1 named after the 8.3 and one of the full name.
Only need full name Example json from the parsed slice_info.config: 3DBenchy_PLA_45m41s.3mf.json

This will print in any scenario now, it will choose the correct filament type, then choose the color that is closest to the hex color value. 

With that,  we won't have to worry if it will choose the right one.  

I plan to do an override for the scenario of starting a print that's been on your machine for awhile but want to print with another color. For now it will automatically choose the right type of filament but choose the closest match to the hex color. So if you have, dark blue, black, grey and light blue in the ams, and let's say, your original print used blue, it would choose the closest hex color to that blue, but if you wanted it to choose black, I'll need to add the manual override. For now it's starting prints wonderfully. 

Also I haven't fixed it yet, but the png files aren't loading because the route isn't right.
It's looking for the png inside plugin folder but I seen it was inside the data folder, also I see that it's creating a file named thumbs and not a folder named thumbs so it's writing data to a file and not adding pngs to a folder. 




Here's a breakdown:
 * Profile Slots (Slicer Side):
   * In Bambu Studio/OrcaSlicer, you define up to 16 "slots" for filaments (4 per potential AMS unit). These are the slots you assign to objects or parts of objects when preparing your print. Think of these as logical filament assignments within your project.
   * When you select a color for an object, you're choosing one of these 16 profile slots.
   * The .3mf file stores which object/part uses which of these 16 profile slots.
 * Physical Trays (Printer Side):
   * The printer detects which physical AMS units are connected and what filament is in each of the 4 slots within those units.
   * Each physical AMS unit has an AMS Index (0 for the first, 1 for the second, etc., based on the daisy-chain order).
   * Each slot within a physical AMS unit has a Slot Index (0 for slot 1, 1 for slot 2, 2 for slot 3, 3 for slot 4).
   * These are combined into a tray_id string in the format "ams_index:slot_index". Examples: "0:0" is AMS #1, Slot 1; "1:3" is AMS #2, Slot 4.
 * The Print Command (start_print) and Mapping:
   When you send an MQTT command to start a print (typically using a command like start_print within a larger print or command structure sent to the /device/{printer_serial_number}/request topic), the payload includes several parameters, including:
   * file: The path to the .3mf file on the printer's SD card or storage.
   * sequence_id: A unique ID for this command (important for tracking).
   * plate_idx: Which plate from the .3mf to print (if there are multiple).
   * And crucially, a tray_mapping.
 * The tray_mapping Parameter - The Key:
   This is a list (array) of exactly 16 entries. Each entry in this list corresponds to one of the 16 potential profile slots from your slicer project (Index 0 of the list corresponds to Profile Slot 1, Index 1 corresponds to Profile Slot 2, ..., Index 15 corresponds to Profile Slot 16).
   The value at each position in the tray_mapping list tells the printer which physical tray_id should be used whenever the .3mf file references that specific profile slot.
 * The -1 Value Explained:
   * If a profile slot in your slicer project (e.g., Profile Slot 5) was not used at all on the plate you are printing (meaning no object or part is assigned to that color/slot), the entry in the tray_mapping list corresponding to that profile slot (Index 4, since it's 0-based) will have the value "-1".
   * Meaning of -1: It signifies "this profile slot is unused for this print/plate, the printer does not need to assign or prepare a physical filament for this slot." The printer will ignore this mapping entry and won't try to use a physical spool for that unused profile slot.
 * Example tray_mapping:
   Let's say your slicer project uses:
   * Profile Slot 1 for Red PLA
   * Profile Slot 5 for Blue ABS
   * Profile Slot 6 for White PETG
   * All other profile slots (2, 3, 4, 7 through 16) are unused on this specific plate.
   And let's say your printer currently has:
   * Physical AMS #1 (index 0), Slot 1 (index 0) loaded with Red PLA (tray_id: "0:0")
   * Physical AMS #2 (index 1), Slot 3 (index 2) loaded with Blue ABS (tray_id: "1:2")
   * Physical AMS #1 (index 0), Slot 4 (index 3) loaded with White PETG (tray_id: "0:3")
   The tray_mapping parameter in your start_print command would look something like this:
   "tray_mapping": [
  "0:0",   // Profile Slot 1 uses physical tray "0:0" (AMS #1, Slot 1)
  "-1",    // Profile Slot 2 is unused
  "-1",    // Profile Slot 3 is unused
  "-1",    // Profile Slot 4 is unused
  "1:2",   // Profile Slot 5 uses physical tray "1:2" (AMS #2, Slot 3)
  "0:3",   // Profile Slot 6 uses physical tray "0:3" (AMS #1, Slot 4)
  "-1",    // Profile Slot 7 is unused
  // ... all the way down to the 16th entry ...
  "-1"     // Profile Slot 16 is unused
]

   The printer reads this map. When the .3mf tells it "print this part using Profile Slot 1", the printer looks at tray_mapping[0], sees "0:0", and knows to use the filament from physical AMS #1, Slot 1. If the .3mf references Profile Slot 2, the printer looks at tray_mapping[1], sees "-1", and does nothing because that slot isn't needed.
 * The Rest of How it Works (Related Concepts):
   * Filament Properties Matching: The printer might check if the filament type, color, and even brand (if known) of the physically loaded filament in the assigned tray_id matches what the .3mf expects for that profile slot. If there's a mismatch or an empty slot where a filament is required (tray_mapping is not "-1" but the physical tray is empty or wrong), the printer will often pause, show an error, or require user intervention.

jneilliii replied:
> Also I haven't fixed it yet, but the png files aren't loading because the route isn't right.
> It's looking for the png inside plugin folder but I seen it was inside the data folder, also I see that it's creating a file named thumbs and not a folder named thumbs so it's writing data to a file and not adding pngs to a folder.

I already had this working for me. It goes into the plugin data folder and the jinja templating is linking through the plugin's route hook via url.

CodeMasterCody3D replied:
I deleted the file named thumbs and it fixed the problem. It's now generating the png files.

CodeMasterCody3D replied:
I recently updated the repo, I'll do pull request and see if you like it. It's the full ams capability, I haven't added the manual override yet. Maybe you could help,  with my lack of knowledge and the terminology,  it's hard for me to find everything I'm trying to do,  then AI gets confused as well.

If you could give me a prompt I could give AI to send me in the right direction I'd be able to add the manual override.  

I tried im at a loss and I don't know what to search to find what I need to know. 

I basically need the UI that you have made with 1 additional thing next to the use ams check box, with a manual override function that would have a drop down,  or we could utilize your existing UI elements for ams. 

I'm sorry I don't know all the terminologies. 

I would need a hook to always reset the manual override to be off after every print to ensure consistency for 1 click printing.  

The override would let you choose same type of filament only,  and warn you if there are no other filaments available.

CodeMasterCody3D replied:
![20250517_010444.jpg](https://github.com/user-attachments/assets/5d5f9b42-f99f-4b86-adb8-449eb20f6f51)

I printed 3 different types of filaments. 
1 click from orca, it will auto select filaments, and choose the closest color with that filament types hex color. 

I made a repo awhile back for manual filament swaps, so you don't have to keep track on a klipper printer. I used an ender 3 pro.

I took the idea i made in that repo and used it with this method.

I plan to build an ams system for klipper, i plan to make it easy like bambu ams. Auto selections and all. 

I'm low on funds so I'm not building an amsv right now.

jneilliii replied:
> I basically need the UI that you have made with 1 additional thing next to the use ams check box, with a manual override function that would have a drop down, or we could utilize your existing UI elements for ams.

I'm going to be working on this directly myself, because I already know the current method of mapping AMS filament during print start is not ideal. the start print dialog that pops up when you start a sd print you can move the spools around on the bottom, and double-click to disable. If you drag them around to re-order that is how they will be reassigned to the print start command. My goal is to use the extracted json data info from the uploaded 3mf file and add that as additional metadata to the sd card file (when that becomes possible to do hopefully in OctoPrint 1.12.0, or not use virtual SD card in ideal world when that's available). This extraction of data is already in the latest rc version, and if you look at your .metadata.json file in OctoPrint's upload folder you will see that. for example, plate data, filament colors and ids are already there for reference as shown below. 

```
  "Alien_Cutie_0821003814.gcode.3mf": {
    "hash": "6bda62aa8a172fe5e5dc11a90a9b1d678643eba7",
    "plate_data": {
      "bbox_all": [
        79.02494899999999,
        80.086798,
        101.229964,
        95.089154
      ],
      "bbox_objects": [
        {
          "area": 108.21379089355469,
          "bbox": [
            79.02494899999999,
            80.086798,
            101.229964,
            95.089154
          ],
          "id": 81,
          "layer_height": 0.20000000298023224,
          "name": "Alien_Cutie_0821003814.stl"
        }
      ],
      "bed_type": "cool_plate",
      "filament_colors": [
        "#F98C36"
      ],
      "filament_ids": [
        0
      ],
      "first_extruder": 0,
      "is_seq_print": false,
      "nozzle_diameter": 0.4000000059604645,
      "version": 2
    }
  },
```

jneilliii replied:
The goal is to mimic the OrcaSlicer/BambuStudio start print dialog that gives you the color/type in the file, and below it a drop down to select the loaded AMS filaments.

CodeMasterCody3D replied:
Ahh, I didn't know if I was able to use the .metadata.json in gear it would mess up octoprints view of the files. Upon closer inspection I see that I can just dump the data I am extracting to the .metadata.json file.

I just tested your most recent changes to your repo, I have petg and pla set in the print but it decided to choose my ASA I have in the AMS. 

I'm going to change the code for mine to place the information in the .metadata.json

So it's all consolidated.

jneilliii replied:
yeah, you can use set_additional_metadata, that's how I store the thumbnail url, and the other stuff that's in the Metadata/plate_1.json file inside the 3mf. You can see it's usage in the block here:

https://github.com/jneilliii/OctoPrint-BambuPrinter/blob/rc/octoprint_bambu_printer/bambu_print_plugin.py#L197-L216

CodeMasterCody3D replied:
I have my part in, I'll will be fixing it tomorrow since I'm off work. 

If you don't mind, what was it that we was working on, which part was you wanting to focus on?

jneilliii replied:
This is where I'm at with your changes. 

https://github.com/jneilliii/OctoPrint-BambuPrinter/pull/83

CodeMasterCody3D committed:
Update bambu_virtual_printer.py

I added time.sleep(5) back to the  
```sh  
 @gcode_executor.register("M524")
    def _cancel_print(self):
```

Reason for this is, if the print gets canceled it will go to operational then immediately start print again then cancels it again if it has time.sleep(0.05), It's harmless but the time.sleep(5) ensures that it has enough time to go though all of It's procedures before it uses the self.remove_project_selection()

I added   

```sh
       time.sleep(5)
        self.remove_project_selection()
```
To the 

```sh
def finalize_print_job(self):
```
After the  self.change_state(self._state_idle)
Reason for this is if you finish a print and try and start a new one it will be stuck and will not start a print from orcaslicer.



```sh
    def finalize_print_job(self):
        if self.current_print_job is not None:
            self.report_print_job_status()
            self.report_print_finished()
            self.current_print_job = None
            self.remove_project_selection()
            self.report_print_job_status()
        self.change_state(self._state_idle)
        time.sleep(5)
        self.remove_project_selection()
```

Changed back to the dosname inside 

```sh
     def _send_file_selected_message(self): 
          if self.selected_file is None: 
              return 
 
 
         self.sendIO(f"File opened: {self.selected_file.dosname} Size: {self.selected_file.size}")
```

CodeMasterCody3D reviewed (commented):
I added time.sleep(5) back to the  
```sh  
 @gcode_executor.register("M524")
    def _cancel_print(self):
```

Reason for this is, if the print gets canceled it will go to operational then immediately start print again then cancels it again if it has time.sleep(0.05), It's harmless but the time.sleep(5) ensures that it has enough time to go though all of It's procedures before it uses the self.remove_project_selection()

I added   

```sh
       time.sleep(5)
        self.remove_project_selection()
```
To the 

```sh
def finalize_print_job(self):
```
After the  self.change_state(self._state_idle)
Reason for this is if you finish a print and try and start a new one it will be stuck and will not start a print from orcaslicer.



```sh
    def finalize_print_job(self):
        if self.current_print_job is not None:
            self.report_print_job_status()
            self.report_print_finished()
            self.current_print_job = None
            self.remove_project_selection()
            self.report_print_job_status()
        self.change_state(self._state_idle)
        time.sleep(5)
        self.remove_project_selection()
```

Changed back to the dosname inside 

```sh
     def _send_file_selected_message(self): 
          if self.selected_file is None: 
              return 
 
 
         self.sendIO(f"File opened: {self.selected_file.dosname} Size: {self.selected_file.size}")
```

jneilliii commented:
> After the self.change_state(self._state_idle) Reason for this is if you finish a print and try and start a new one it will be stuck and will not start a print from orcaslicer.
> 
> ```shell
>     def finalize_print_job(self):
>         if self.current_print_job is not None:
>             self.report_print_job_status()
>             self.report_print_finished()
>             self.current_print_job = None
>             self.remove_project_selection()
>             self.report_print_job_status()
>         self.change_state(self._state_idle)
>         time.sleep(5)
>         self.remove_project_selection()
> ```

This confuses me a bit. I feel like duplicating calls to remove_project_selection is unnecessary from a logic perspective. 

I feel like if the sleep is necessary it should be inside the if, and I question if 5 seconds is too long. Feel like it should be max 1 second. 

```
    def finalize_print_job(self):
        if self.current_print_job is not None:
            self.report_print_job_status()
            self.report_print_finished()
            self.current_print_job = None
            time.sleep(1)
            self.remove_project_selection()
            self.report_print_job_status()
        self.change_state(self._state_idle)
```

CodeMasterCody3D commented:
What I think is happening is the 5 secs makes it time out and doesn't let it continue to loop and make it repeat the print cancel pattern. When in use on the  gcode cancel. 



I assume that's why the 5 sec sleep on the finalize_print_job also makes it work too i assume..... so that it doesn't get stuck in some weird "state" or "loop" or something.  


I meant to explain it before.  


Maybe that information will be enough to help you diagnose it properly. 

I have an idea.  Imma remove the (remove_project_selection) and see if just having the time sleep 5, keeps the same effect. 


I'll be back.


Testing by removing the remove_project_selection on the gcode cancel and the finalize_print_job
